packages/utils/src/cache/cache-indexeddb.ts

```diff
@@ -0,0 +1,168 @@
+import { openDB } from 'idb';
+import type { DBSchema, IDBPDatabase } from 'idb';
+import type { CacheMap } from './types';
+import type { CacheItem } from './interfaces/CacheItem';
+import type { CacheStorage } from './interfaces/CacheStorage';
+import { PREFIX } from './constants.js';
+
+interface IndexedDBDatabase extends DBSchema {
+  [key: string]: {
+    key: string;
+    value: CacheItem;
+  }
+}
+
+/**
+ * Stores data in indexedDB for use across multiple sessions.
+ */
+export class CacheIndexedDBStorage implements CacheStorage {
+  /**
+   * Prefix for database name, and store name
+   * to avoid database to clash with the workspace
+   */
+  protected prefixKey = '';
+
+  /**
+   * A connection to a indexedDB database, use for open transaction or idb api
+   */
+  private db: IDBPDatabase<IndexedDBDatabase> | undefined;
+
+  /**
+   * Database Version.
+   */
+  private version = 1;
+
+  /**
+   * Internal cache object
+   */
+  protected cache: CacheMap | null | undefined;
+
+  /**
+   * Cache ready to use
+   */
+  protected ready: Promise<boolean> | null = null;
+
+  /**
+   * Constructor
+   * @param name for database name and store name
+   */
+  constructor (name: string) {
+    this.prefixKey = PREFIX + (name || '');
+    void this.open();
+  }
+
+  /**
+   * Set a item against a key to this storage
+   * @param key item key
+   * @param value item value
+   * @returns {void}
+   */
+  public async setItem (key: string, value: CacheItem): Promise<void> {
+    await this.ready;
+    const item = { ...value, key };
+    this.cache?.set(key, item);
+    await this.db?.put(this.prefixKey as never, item, key);
+  }
+
+  /**
+   * Returns a item in this storage that matched by the key.
+   * @param key item key
+   * @returns cache item or null
+   */
+  public async getItem (key: string): Promise<CacheItem | null> {
+    await this.ready;
+    return this.cache?.get(key) || null;
+  }
+
+  /**
+   * Remove a item against a key to this storage
+   * @param key item key
+   * @returns {void}
+   */
+  public async removeItem (key: string): Promise<void> {
+    await this.ready;
+    this.cache?.delete(key);
+    await this.db?.delete(this.prefixKey as never, key);
+  }
+
+  /**
+   * Clear all item in this storage
+   * @returns {void}
+   */
+  public async clear (): Promise<void> {
+    await this.ready;
+    this.cache?.clear();
+    await this.db?.clear(this.prefixKey as never);
```

Comment:
Interesting choice of casting key to `never` here. Is there better ways to do this, from looking at `idb` dts they do offer some types.

Reply:
Updated to remove type casting.
I and Goh propose an idea to use literal type ``type StoreName = `[${StoragePrefix.DEFAULT}][${string}]` ``, and cast it into store type of idb (thrid party indexeddb).